x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -188,33 +206,40 @@ protected void readInternal(final byte[] buffer, final int offset, final int len
                     if (e instanceof AlreadyClosedException || (e.getCause() != null && e.getCause() instanceof AlreadyClosedException)) {
                         try {
                             // cache file was evicted during the range fetching, read bytes directly from source
-                            bytesRead += readDirectly(pos, pos + len, buffer, off);
+                            bytesRead = readDirectly(pos, pos + len, buffer, off);
+                            stats.addDirectBytesRead(bytesRead);
                             continue;
                         } catch (Exception inner) {
                             e.addSuppressed(inner);
                         }
                     }
                     throw new IOException("Fail to read data from cache", e);
 
+                } finally {
+                    totalBytesRead += bytesRead;
                 }
             }
-            assert bytesRead == length : "partial read operation, read [" + bytesRead + "] bytes of [" + length + "]";
+            assert totalBytesRead == length : "partial read operation, read [" + totalBytesRead + "] bytes of [" + length + "]";
+            stats.incrementBytesRead(lastReadPosition, position, totalBytesRead);
+            lastReadPosition = position + totalBytesRead;
         }
 
         int readCacheFile(FileChannel fc, long end, long position, byte[] buffer, int offset, long length) throws IOException {
             assert assertFileChannelOpen(fc);
-            return Channels.readFromFileChannel(fc, position, buffer, offset, Math.toIntExact(Math.min(length, end - position)));
+            int bytesRead = Channels.readFromFileChannel(fc, position, buffer, offset, Math.toIntExact(Math.min(length, end - position)));
+            stats.addCachedBytesRead(bytesRead);
+            return bytesRead;
         }
 
         @SuppressForbidden(reason = "Use positional writes on purpose")
         void writeCacheFile(FileChannel fc, long start, long end) throws IOException {
             assert assertFileChannelOpen(fc);
             final byte[] copyBuffer = new byte[Math.toIntExact(Math.min(COPY_BUFFER_SIZE, end - start))];
+            int bytesCopied = 0;
             try (IndexInput input = in.openInput(cacheFileReference.getFileName(), ioContext)) {
```

Comment:
Also, I think we should add a `TRACE`-level log recording the specific range we're copying here (and in `readDirectly`) because I think there will be situations where we need to see the details of the access pattern.

Reply:
Added in 208b14f3161859c6f6d31da70e9499165c12f534